Added qc classes for PID and Clusters

tklemenz opened the pull request:
- added qc/PID, qc/TrackCut, qc/Clusters classes
- added Helper functions in o2::tpc::qc::helpers
- the o2::tpc::painter::draw functions now return the canvas

wiechula commented:
Hi @tklemenz ,

all macros need to be compilable. How to do this, you can have a look e.g. here:
https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/TPC/calibration/CMakeLists.txt
for the CMakeLists.txt and here:
https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/TPC/calibration/macro/drawNoiseAndPedestal.C
for what to include in the macro around the inlcudes:
```c++
#if !defined(__CLING__) || defined(__ROOTCLING__)
...
#endif
```
Cheers,
Jens

wiechula commented on Detectors/TPC/qc/src/Helpers.cxx:
```diff
@@ -0,0 +1,50 @@
+void helpers::setStyleHistogram1D(TH1& histo)
+{
+  histo.SetStats(1);
+}
```
It might be nice to have a function which takes a vector of histograms. This will be used in many places I guess. Same for 2D.

tklemenz replied:
Only managed to implement a function that takes std::vector<TH2F>* instead of std::vector<TH2>*. I will figure out how o generalize this or implement a function for every single TH2 variant.

tklemenz replied:
Same of course for TH1.

tklemenz replied:
just noticed that formatting went wrong in my older comment. I implemented the function for a std::vector* of TH2F and TH1F but not generalized for all TH1 and TH2 variants.

wiechula commented on Detectors/TPC/qc/src/PID.cxx:
```diff
@@ -8,50 +8,108 @@
+
+  for (auto& hist : mHist2D) {
+    helpers::setStyleHistogram2D(hist);
+  }
```
As written above, perhaps the helpers can have a function which takes the vector.

tklemenz replied:
see above

tklemenz commented on Detectors/TPC/base/include/TPCBase/Painter.h:
```diff
@@ -17,6 +17,7 @@
 ///
 
 #include "DataFormatsTPC/Defs.h"
+#include "TCanvas.h"
```
done

tklemenz commented on Detectors/TPC/qc/CMakeLists.txt:
```diff
@@ -1,24 +1,53 @@
+                       src/Helpers.cxx
+		       src/TrackCuts.cxx
+		       src/Clusters.cxx
                PUBLIC_LINK_LIBRARIES O2::TPCBase)
```
done

wiechula replied:
I think now you used a mix of spaces and tabs. Can you convert this all to spaces.

tklemenz commented on Detectors/TPC/qc/include/TPCQC/Clusters.h:
```diff
@@ -0,0 +1,73 @@
+
+  void dumpToFile(std::string filename);
+
+  TCanvas* paint(const CalDet<float>& calDet);
```
gone

tklemenz commented on Detectors/TPC/qc/include/TPCQC/Clusters.h:
```diff
@@ -0,0 +1,73 @@
+  CalPad mQTot{"Q_Tot"};
+  CalPad mSigmaTime{"Sigma_Time"};
+  CalPad mSigmaPad{"Sigma_Pad"};
+  CalPad mTimeBin{"TimeBin"};
```
done

tklemenz commented on Detectors/TPC/qc/include/TPCQC/Helpers.h:
```diff
@@ -0,0 +1,49 @@
+
+//root includes
+#include "TH1F.h"
+#include "TH2F.h"
```
done

tklemenz commented on Detectors/TPC/qc/include/TPCQC/PID.h:
```diff
@@ -10,18 +10,28 @@
 #include "TH1F.h"
+#include "TH2F.h"
+#include "TFile.h"
+#include "TCanvas.h"
```
done

tklemenz commented on Detectors/TPC/qc/include/TPCQC/PID.h:
```diff
@@ -10,18 +10,28 @@
+/// from TrackTPC objects and store it in histograms
+///
+/// origin: TPC
+/// @author Thomas Klemenz, thomas.klemenz@tum.de
```
done

tklemenz commented on Detectors/TPC/qc/include/TPCQC/PID.h:
```diff
@@ -10,18 +10,28 @@
+#include "TCanvas.h"
 //o2 includes
 #include "DataFormatsTPC/Defs.h"
 #include "DataFormatsTPC/TrackTPC.h"
```
TrackTPC as forward declaration does not work somehow.

wiechula replied:
See my comment in TrackCuts.

tklemenz commented on Detectors/TPC/qc/include/TPCQC/PID.h:
```diff
@@ -30,15 +40,14 @@ namespace tpc
+  PID() = default;
 
+  /// bool extracts intormation from track and fills it to histograms
+  // \return true if information can be extracted and filled to histograms
```
done

tklemenz commented on Detectors/TPC/qc/src/TrackCuts.cxx:
```diff
@@ -0,0 +1,48 @@
+
+TrackCuts::TrackCuts() : mPMin(-999),
+                         mPMax(-999),
+                         mNClusMin(-999)
```
Default constructor now initializes with the suggested values.

tklemenz commented on Detectors/TPC/qc/include/TPCQC/TrackCuts.h:
```diff
@@ -0,0 +1,54 @@
+#ifndef AliceO2_TPC_TRACKCUTS_H
+#define AliceO2_TPC_TRACKCUTS_H
+
+#include "DataFormatsTPC/TrackTPC.h"
```
TrackTPC as forward declaration also here does not work.

wiechula replied:
can you please move the 
```
#include "DataFormatsTPC/TrackTPC.h"
```
to the cxx and add
```
#include "Rtypes.h"
```
instead.
then see below.

tklemenz commented on Detectors/TPC/qc/include/TPCQC/TrackCuts.h:
```diff
@@ -0,0 +1,54 @@
+ private:
+  float mPMin;     // min momentum allowed
+  float mPMax;     // max momentum allowed
+  float mNClusMin; // min number of clusters in track allowed
```
Default initialization now with suggested values.

tklemenz commented on Detectors/TPC/qc/macro/runClusters.C:
```diff
@@ -0,0 +1,46 @@
+void runClusters(std::string_view fileName = "tpc-native-clusters.root", std::string_view outputFile = "ClusterQC", const size_t maxNClusters = 0)
```
done

tklemenz commented on Detectors/TPC/qc/macro/runPID.C:
```diff
@@ -0,0 +1,61 @@
+using namespace o2::tpc;
```
done

tklemenz replied:
On my local machine now everything compiles including the macros. I can also compile on the fly when calling the macro with root. 
Here the checks fail because TPCQC/Clusters.h and TPCQC/PID.h are not found in the corresponding macro. Any idea why?

wiechula replied:
Hmm, funny. I think this should work. We can check in separately.

tklemenz commented on Detectors/TPC/qc/src/Clusters.cxx:
```diff
@@ -0,0 +1,182 @@
+  o2::tpc::painter::draw(mTimeBin);
+}
+
+TCanvas* Clusters::paint(const CalDet<float>& calDet)
```
removed

tklemenz commented on Detectors/TPC/qc/src/Clusters.cxx:
```diff
@@ -0,0 +1,182 @@
+  f->WriteObject(Clusters::paint(mQTot), "Q_Tot");
+  f->WriteObject(Clusters::paint(mSigmaTime), "Sigma_Time");
+  f->WriteObject(Clusters::paint(mSigmaPad), "Sigma_Pad");
+  f->WriteObject(Clusters::paint(mTimeBin), "TimeBin");
```
done

wiechula commented on Detectors/TPC/qc/include/TPCQC/TrackCuts.h:
```diff
@@ -0,0 +1,58 @@
+namespace o2
+{
+namespace tpc
+{
```
add
```
class TrackTPC;
```
here.

This should work. At least it did for me.

tklemenz replied:
Thanks, this works.

wiechula commented on Detectors/TPC/qc/CMakeLists.txt:
```diff
@@ -1,24 +1,55 @@
+                       src/Helpers.cxx
+                       src/TrackCuts.cxx
+                       src/Clusters.cxx
                PUBLIC_LINK_LIBRARIES O2::TPCBase)
```
can you please add O2::DataFormatsTPC to the PUBLIC_LINK_LIBRARIES this should fix the compilation error

wiechula commented on Detectors/TPC/qc/src/Helpers.cxx:
```diff
@@ -0,0 +1,73 @@
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include <assert.h>
```
Please use casser instead of assert.h

tklemenz replied:
done

tklemenz commented:
I just noticed that not only the O2 but also the Quality Control package is built in the test. Now it fails because the `#include "DataFormatsTPC/TrackTPC.h"` is missing from `QualityControl/Modules/TPC/include/TPC/PID.h`.
I managed to reproduce the error and could fix it by including the corresponding header. This I think has nothing to do with the pull request and should be fixed somehow. What do you think, @wiechula? 
Should I do a pull request for this in the Quality Control?

wiechula commented:
Hi @tklemenz ,

since the TrackTPC is not used in the header, but only in the cxx, please add it to PID.cxx not PID.h!
After that, a PR would be good, yes.

Thanks,
Jens

tklemenz commented:
Ok, now in O2 I put the forward declaration of the TrackTPC in the PID.h and the include in the PID.cxx. In QC I put the include of the TrackTPC in the PID.cxx and did a pull request.
I thought we could maybe avoid the pull request in QC but I hope that goes quickly now.

Cheers, 
Thomas

wiechula commented on Detectors/TPC/qc/include/TPCQC/TrackCuts.h:
```diff
@@ -0,0 +1,61 @@
+ private:
+  float mPMin{0};     // min momentum allowed
+  float mPMax{1e10};  // max momentum allowed
+  float mNClusMin{0}; // min number of clusters in track allowed
```
please format comments with doxygen, so e.g.
```c++
  float mPMin{0};     ///< min momentum allowed
```

tklemenz replied:
WIP

wiechula commented on Detectors/TPC/qc/include/TPCQC/Clusters.h:
```diff
@@ -0,0 +1,69 @@
+
+  void analyse();
+
+  void dumpToFile(std::string filename);
```
I would prefer `std::string_view` and `#include <string_view>`.

tklemenz replied:
What I wanted to achieve is that the user can give a name for the output to the runClusters.C so that they can either type `name.root` or simply `name`. Then they receive a `name.root` file containing the data members of the CalPad objects and another file called `name_canvas.root` containing the canvases that were drawn by the painter. 
As far as I see this was easy with std:: string objects but is not possible with std::string_view objects. Imho std::string_view objects can not be manipulated. Also functions like ends_with() that would simplify the code e.g. `if (filename.find(".root") != std::string::npos) {...` only work in C++20. 
I would suggest to stick with std::string objects or am I missing something?

wiechula replied:
See below in Clusters.cxx

wiechula commented on Detectors/TPC/qc/src/Clusters.cxx:
```diff
@@ -0,0 +1,95 @@
+{
+  if (filename.find(".root") != std::string::npos) {
+    filename.resize(filename.size() - 5);
+  }
```
If you used string_views they can easily be manipulated. Please check the documentation.

tklemenz replied:
see above

wiechula replied:
Could work like this:
```c++
std::string_view filename = "test.root";
const auto rootpos = filename.rfind(".root");
if (rootpos != std::string_view::npos) {
  filename=filename.substr(0, rootpos);
}

std::string outFileBase(filename);

TFile::Open((outFileBase + "_canvas.root").c_str(), "recreate");
```
But as you like since this is totally unimportant...

wiechula commented on Detectors/TPC/qc/src/Clusters.cxx:
```diff
@@ -0,0 +1,95 @@
+  g->WriteObject(&mSigmaTime.getData(), "Sigma_Time");
+  g->WriteObject(&mSigmaPad.getData(), "Sigma_Pad");
+  g->WriteObject(&mTimeBin.getData(), "Time_Bin");
+  g->Close();
```
perhaps you could add an option to dumpToFile which allows choosing what to save.

tklemenz replied:
WIP

wiechula commented on Detectors/TPC/qc/src/Clusters.cxx:
```diff
@@ -0,0 +1,95 @@
+  f->WriteObject(o2::tpc::painter::draw(mSigmaTime), "Sigma_Time");
+  f->WriteObject(o2::tpc::painter::draw(mSigmaPad), "Sigma_Pad");
+  f->WriteObject(o2::tpc::painter::draw(mTimeBin), "Time_Bin");
+  f->Close();
```
Why not use the name of the object? E.g.
```c++
f->WriteObject(o2::tpc::painter::draw(mTimeBin), mTimeBin.getName());
```
and so on. Same below.

tklemenz replied:
done

wiechula commented on Detectors/TPC/qc/src/Helpers.cxx:
```diff
@@ -0,0 +1,73 @@
+void helpers::setStyleHistogram1D(std::vector<TH1F>* histos)
+{
+  for (auto& hist : *histos) {
+    hist.SetStats(1);
```
Why not call `setStyleHistogram1D`? Otherwise you will get different behaviour in both cases or need to maintain two functions ... Same below for the 2D case.

tklemenz replied:
This make a lot of sense. 
done

wiechula commented on Detectors/TPC/qc/src/PID.cxx:
```diff
@@ -8,50 +8,99 @@
+  f->WriteObject(&mHist2D[1], "dEdxVstgl");
+  f->WriteObject(&mHist2D[2], "dEdxVsnclusters");
+  f->WriteObject(&mHist2D[3], "dEdxVsp");
+  f->Close();
```
Also here, why not
```c++
for (auto& hist : mHist1D) {
  f->WriteObject(&hist, hist.GetName());
}
```
also for 2D. Otherwise you would need to modify the function whenever you add histograms.

Or even better if this works for vectors of histograms would be to write the whole collection:
```c++
f->Write(&mHist1D, "Hist1D");
f->Write(&mHist2D, "Hist2D");
```
To be tried out.

tklemenz replied:
For now I am going with writing the single objects to file in a loop as you suggested. I tried writing the whole vectors and they are written to file but 

1. the root browser can not properly handle the vectors so there is no histogram to look at
2. I did not yet manage to draw the histograms from the root shell but that may be a user problem ;)

I will further investigate that.

wiechula replied:
Ok, then lets keep it with the individual histograms written as it is.

tklemenz commented:
Hi @wiechula, 
I just pushed this for Stefan to have the latest version. Not all your comments are implemented, yet (but most of them are). 
Cheers, 
Thomas

tklemenz commented on Detectors/TPC/qc/include/TPCQC/Clusters.h:
```diff
@@ -0,0 +1,69 @@
+
+//o2 includes
+#include "DataFormatsTPC/Defs.h"
+#include "DataFormatsTPC/ClusterNative.h"
```
done

tklemenz commented on Detectors/TPC/qc/include/TPCQC/Clusters.h:
```diff
@@ -0,0 +1,69 @@
+ public:
+  Clusters() = default;
+
+  bool processCluster(o2::tpc::ClusterNative const& cluster, o2::tpc::Sector const sector, const int row);
```
I did it the other way around and did the others as "const type name" because it was done like this everywhere else.

tklemenz commented on Detectors/TPC/qc/include/TPCQC/Helpers.h:
```diff
@@ -0,0 +1,54 @@
+
+#ifndef AliceO2_TPC_HELPERS_H
+#define AliceO2_TPC_HELPERS_H
+
```
done

tklemenz commented on Detectors/TPC/qc/include/TPCQC/Helpers.h:
```diff
@@ -0,0 +1,54 @@
+void setStyleHistogram1D(TH1& histo);
+
+/// Set nice style for vector of 1D histograms
+void setStyleHistogram1D(std::vector<TH1F>* histos);
```
This works with forward declarations. :)

tklemenz commented on Detectors/TPC/qc/include/TPCQC/PID.h:
```diff
@@ -47,16 +56,25 @@ class PID
   void resetHistograms();
 
+  /// Dump results to a file
+  void dumpToFile(const std::string filename);
```
Imho string_view does not work the way I need it to. (see below)

tklemenz commented on Detectors/TPC/qc/include/TPCQC/PID.h:
```diff
@@ -10,35 +10,44 @@
-/// @author Jens Wiechula, Jens.Wiechula@ikf.uni-frankfurt.de
+/// @author Thomas Klemenz, thomas.klemenz@tum.de
 ///
 
```
done

tklemenz commented on Detectors/TPC/qc/include/TPCQC/PID.h:
```diff
@@ -10,35 +10,44 @@
-#ifndef AliceO2_TPC_ROC_H
-#define AliceO2_TPC_ROC_H
+#ifndef AliceO2_TPC_PID_H
+#define AliceO2_TPC_PID_H
```
done

tklemenz commented:
The final problem I had with string_view was that in the runPID.C when the string_view object was given to the dumpToFile function there was an error that TString could not be converted to string_view. However, there was no TString involved anywhere. Also Stefan was confused and did not know how to solve this. Otherwise I got everything working but the code was not shorter than with using std::string.

For the cluster class I could implement that but in principle the PID class should have the same.
I would leave it like this for the moment so the code can be used now and change this when further changes are implemented e.g. proper tests for the classes and helper functions.

wiechula commented:
Sure, not important as I wrote. Please go ahead with the PR.

tklemenz commented:
Imho PR #279 in Quality Control fixed the `invalid application of 'sizeof' to incomplete type 'std::vector<o2::tpc::TrackTPC>::value_type {aka o2::tpc::TrackTPC}'` error that comes up in `build/O2/gpu`. Not sure why this still comes up.

wiechula commented:
Hi @tklemenz ,

I think the problem is that a new release of QC would be needed. It seems that not the master is used but v0.19.5. @Barthelemy what is the way to go here? How often is QC tagged an then used in the tests? We could also modify our code in O2 again that the test will work, but I think this is not the proper way.

Cheers,
Jens

Barthelemy commented:
Hi Jens, 

You are correct. The tests will use the version defined in the recipe, not master. 

We do not have a fixed schedule for releases in QC. We do it when there are necessary changes (like here) or when we have a growing number of new features that we want to bring to the users. 

I will release 0.19.6, then update the recipe and then your tests here should be working. If you need new releases in the future, you can ping me and I will do it. 

This is one of the reasons I was managing the detectors modules independently from the framework and from each other in AMORE. It made it really easy for the detectors to build their own package whenever they needed it. Not sure how this would be compatible with the O2 philosophy. 

Cheers,
Barth

wiechula commented:
Hi @Barthelemy ,

thanks a lot for your extremely fast response!
I think you are right. The design of AMORE more reflected the needs of the detectors. Perhaps we will need to re-discuss this at some point also for o2.
For the moment we will ping you in case we need release as you suggested.

Thanks again.
Cheers,
Jens

Barthelemy commented:
QC 0.19.6 is released and set in the recipe.

tklemenz commented:
Thanks a lot, Barth!
I rebased on dev and pushed again to trigger new checks.

wiechula commented:
Cool, now all checks went through.
@tklemenz , can you remove the [WIP] in the title, then it can go to dev.

tklemenz commented:
@wiechula, I changed the title. Are you notified when something changes on this pull request? If not I'm sorry for not commenting right away when I changed it.

wiechula commented:
Hi @tklemenz ,
nor worry I get the notifications.

wiechula commented:
Hi @davidrohr, @shahor02 ,
can this be merged from your side?
